source/extensions/filters/http/ext_proc/client_impl.cc

```diff
@@ -62,8 +62,9 @@ void ExternalProcessorStreamImpl::send(envoy::service::ext_proc::v3::ProcessingR
 bool ExternalProcessorStreamImpl::close() {
   if (!stream_closed_) {
     ENVOY_LOG(debug, "Closing gRPC stream");
-    if (grpc_side_stream_flow_control_) {
+    if (grpc_side_stream_flow_control_ && !watermark_callbacks_removed_) {
       stream_.removeWatermarkCallbacks();
+      watermark_callbacks_removed_ = true;
```

Comment:
hm, I wonder if we should have a single clean up function for shared code?  having to have a tracking boolean makes me think we have some unnecessary complexity here

Reply:
Yeah, I have thought about this. It is specifically related to ext_proc logic: In most cases,  the stream will be closed before filter is destroyed. So calling this removal in single `ExternalProcessorStreamImpl::close` is good.  

However,  in ext_proc' [observability mode](https://github.com/tyxia/envoy/blob/a960f5ceb1b2812aea59e63bf6180afb8c866b68/source/extensions/filters/http/ext_proc/ext_proc.cc#L401-L404) where stream outlives filter.  In this case, `watermarkcallbacks` should also be removed as soon as filter onDestroy, so that the referenced filter callbacks is not accessible anymore.   And single `ExternalProcessorStreamImpl::close` can not cover this case as it is deferred clsoed.

I actually split this change into https://github.com/envoyproxy/envoy/pull/35298 for easier review and faster roll-out. Its impact and complexity is limited only under `config_->observabilityMode()`. For the normal non-observability mode, this flag is like a no-op.

WDYT?

Reply:
I think it'd still be good to have shared cleanup/detatch code called in both cases, so if there are other things we need to clean up, we don't have the next person who is clueless about the 2 modes miss observability mode cleanup :-)